hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/io/HoodieSparkMergeHandleV2.java

```diff
@@ -0,0 +1,413 @@
+    }
+  }
+
+  @Override
+  public List<WriteStatus> close() {
+    try {
+      if (isClosed()) {
+        // Handle has already been closed
+        return Collections.emptyList();
+      }
+
+      markClosed();
+
+      writtenRecordKeys = null;
+
+      fileWriter.close();
+      fileWriter = null;
+
+      long fileSizeInBytes = storage.getPathInfo(newFilePath).getLength();
+      HoodieWriteStat stat = writeStatus.getStat();
+
+      stat.setTotalWriteBytes(fileSizeInBytes);
+      stat.setFileSizeInBytes(fileSizeInBytes);
+      stat.setNumWrites(recordsWritten);
+      stat.setNumDeletes(recordsDeleted);
+      stat.setNumUpdateWrites(updatedRecordsWritten);
+      stat.setNumInserts(insertRecordsWritten);
+      stat.setTotalWriteErrors(writeStatus.getTotalErrorRecords());
+      RuntimeStats runtimeStats = new RuntimeStats();
+      runtimeStats.setTotalUpsertTime(timer.endTimer());
+      stat.setRuntimeStats(runtimeStats);
+
+      performMergeDataValidationCheck(writeStatus);
+
+      LOG.info(String.format("MergeHandle for partitionPath %s fileID %s, took %d ms.", stat.getPartitionPath(),
+          stat.getFileId(), runtimeStats.getTotalUpsertTime()));
+
+      return Collections.singletonList(writeStatus);
+    } catch (IOException e) {
+      throw new HoodieUpsertException("Failed to close UpdateHandle", e);
+    }
+  }
+
+  public void performMergeDataValidationCheck(WriteStatus writeStatus) {
+    if (!config.isMergeDataValidationCheckEnabled()) {
+      return;
+    }
+
+    long oldNumWrites = 0;
+    try (HoodieFileReader reader = HoodieIOFactory.getIOFactory(hoodieTable.getStorage())
+        .getReaderFactory(this.recordMerger.getRecordType())
+        .getFileReader(config, oldFilePath)) {
+      oldNumWrites = reader.getTotalRecords();
+    } catch (IOException e) {
+      throw new HoodieUpsertException("Failed to check for merge data validation", e);
+    }
+
+    if ((writeStatus.getStat().getNumWrites() + writeStatus.getStat().getNumDeletes()) < oldNumWrites) {
+      throw new HoodieCorruptedDataException(
+          String.format("Record write count decreased for file: %s, Partition Path: %s (%s:%d + %d < %s:%d)",
+              writeStatus.getFileId(), writeStatus.getPartitionPath(),
+              instantTime, writeStatus.getStat().getNumWrites(), writeStatus.getStat().getNumDeletes(),
+              baseFileToMerge.getCommitTime(), oldNumWrites));
+    }
+  }
+
+  public Iterator<List<WriteStatus>> getWriteStatusesAsIterator() {
+    List<WriteStatus> statuses = getWriteStatuses();
+    // TODO(vc): This needs to be revisited
+    if (getPartitionPath() == null) {
+      LOG.info("Upsert Handle has partition path as null {}, {}", getOldFilePath(), statuses);
+    }
+    return Collections.singletonList(statuses).iterator();
+  }
+
+  public StoragePath getOldFilePath() {
+    return oldFilePath;
+  }
+
+  @Override
+  public IOType getIOType() {
```

Comment:
please verify markers etc work well even with the new merge handle (IIRC this method is used in that context)

Reply:
Verified that through the new merge handle based on the file group reader, markers are still created:
```
ethan@MacBook-Pro-3 /p/v/f/6/w/T/s/h1> ls .hoodie/timeline/
20241204211056637.deltacommit.inflight           20241204211108979_20241204211111563.deltacommit  20241204211114600.deltacommit.requested
20241204211056637.deltacommit.requested          20241204211112429.deltacommit.inflight           20241204211114600_20241204211118658.deltacommit
20241204211056637_20241204211108220.deltacommit  20241204211112429.deltacommit.requested          20241204211118741.compaction.inflight
20241204211108979.deltacommit.inflight           20241204211112429_20241204211114352.deltacommit  20241204211118741.compaction.requested
20241204211108979.deltacommit.requested          20241204211114600.deltacommit.inflight           history/
ethan@MacBook-Pro-3 /p/v/f/6/w/T/s/h1> cd /private/var/folders/60/wk8qzx310fd32b2dp7mhzvdc0000gn/T/spark-682128ea-6fde-4a1d-bcad-4f12e427b3c2/h1
ethan@MacBook-Pro-3 /p/v/f/6/w/T/s/h1> ls
93990da3-b7e9-4eaf-97c2-067e7f3d8355-0_0-136-184_20241204211118741.parquet
93990da3-b7e9-4eaf-97c2-067e7f3d8355-0_0-21-36_20241204211056637.parquet
ethan@MacBook-Pro-3 /p/v/f/6/w/T/s/h1> cd .hoodie/
ethan@MacBook-Pro-3 /p/v/f/6/w/T/s/h/.hoodie> ls
hoodie.properties  metadata/          timeline/
ethan@MacBook-Pro-3 /p/v/f/6/w/T/s/h/.hoodie> cd .temp/
ethan@MacBook-Pro-3 /p/v/f/6/w/T/s/h/.h/.temp> ls
20241204211118741/
ethan@MacBook-Pro-3 /p/v/f/6/w/T/s/h/.h/.temp> cd 20241204211118741/
ethan@MacBook-Pro-3 /p/v/f/6/w/T/s/h/.h/.t/20241204211118741> ls
MARKERS.type  MARKERS0
ethan@MacBook-Pro-3 /p/v/f/6/w/T/s/h/.h/.t/20241204211118741> ls -l
total 16
-rw-r--r--@ 1 ethan  staff  21 Dec  4 13:11 MARKERS.type
-rw-r--r--@ 1 ethan  staff  88 Dec  4 13:11 MARKERS0
ethan@MacBook-Pro-3 /p/v/f/6/w/T/s/h/.h/.t/20241204211118741> cat MARKERS0 
93990da3-b7e9-4eaf-97c2-067e7f3d8355-0_0-136-184_20241204211118741.parquet.marker.MERGE

```